src/sys/unix.rs

```diff
@@ -659,6 +665,75 @@ impl SockAddr {
             None
         }
     }
+
+    /// Returns true if this address is an unnamed address from the `AF_UNIX` (AKA `AF_LOCAL`)
+    /// family (for local interprocess communication), false otherwise.
+    pub fn is_unnamed(&self) -> bool {
+        self.as_sockaddr_un()
+            .map(|storage| {
+                self.len() == offset_of_path(storage) as u32
+                    // On some non-linux platforms a zeroed path is returned for unnamed.
+                    // Abstract addresses only exist on Linux.
+                    || (cfg!(not(any(target_os = "linux", target_os = "android")))
+                    && storage.sun_path[0] == 0)
+            })
+            .unwrap_or_default()
+    }
+
+    /// Returns the underlying `sockaddr_un` object if this addres is from the `AF_UNIX` family,
+    /// otherwise returns `None`.
+    pub(crate) fn as_sockaddr_un(&self) -> Option<&libc::sockaddr_un> {
+        self.is_unix().then(|| {
+            // SAFETY: if local, i.e. the `ss_family` field is `AF_UNIX` then storage must be a
+            // `sockaddr_un`.
+            unsafe { &*ptr::addr_of!(self.storage).cast::<libc::sockaddr_un>() }
```

Comment:
why not use `self.as_storage_ptr`? I assume that's why they added this?

Reply:
`as_storage_ptr` is not defined on `redox`.

This expression is what's used in the existing code to get a reference to the appropriate socket address type:

https://github.com/rust-lang/socket2/blob/18a883653752116a10ec78f7d36f8d956e4c2e7c/src/sockaddr.rs#L199

https://github.com/rust-lang/socket2/blob/18a883653752116a10ec78f7d36f8d956e4c2e7c/src/sockaddr.rs#L206

https://github.com/rust-lang/socket2/blob/18a883653752116a10ec78f7d36f8d956e4c2e7c/src/sys/unix.rs#L577

Reply:
hmm but that requires the visibility change that might be there to avoid using stuff on wrong platform.

Reply:
p.s `as_ptr` might work, no?

Reply:
Changed to call `as_ptr` instead of calling `addr_of!` directly.